src/back/spv/writer.rs

```diff
@@ -1114,60 +1129,57 @@ impl Writer {
-                            crate::TypeInner::Scalar { .. } => (
-                                super::instructions::instruction_matrix_times_scalar(
-                                    left_result_type_id,
-                                    id,
-                                    left_id,
-                                    right_id,
-                                ),
-                                left_lookup_ty,
+                            left_lookup_ty,
+                        ),
+                        (Dimension::Matrix, Dimension::Scalar { .. }) => (
+                            super::instructions::instruction_matrix_times_scalar(
+                                left_result_type_id,
+                                id,
+                                left_id,
+                                right_id,
                             ),
-                            crate::TypeInner::Vector { .. } => (
-                                super::instructions::instruction_matrix_times_vector(
-                                    right_result_type_id,
-                                    id,
-                                    left_id,
-                                    right_id,
-                                ),
-                                right_lookup_ty,
+                            left_lookup_ty,
+                        ),
+                        (Dimension::Matrix, Dimension::Vector) => (
+                            super::instructions::instruction_matrix_times_vector(
+                                right_result_type_id,
+                                id,
+                                left_id,
+                                right_id,
                             ),
-                            crate::TypeInner::Matrix { .. } => (
-                                super::instructions::instruction_matrix_times_matrix(
-                                    left_result_type_id,
-                                    id,
-                                    left_id,
-                                    right_id,
-                                ),
-                                left_lookup_ty,
+                            right_lookup_ty,
+                        ),
+                        (Dimension::Matrix, Dimension::Matrix) => (
+                            super::instructions::instruction_matrix_times_matrix(
+                                left_result_type_id,
+                                id,
+                                left_id,
+                                right_id,
                             ),
-                            _ => unreachable!(),
-                        },
-                        crate::TypeInner::Scalar { kind, .. } => match kind {
+                            left_lookup_ty,
+                        ),
+                        (Dimension::Scalar { kind, .. }, Dimension::Scalar { .. }) => match kind {
```

Comment:
can we match more aggressively here? i.e.
```rust
(Dimension::Scalar { kind: crate::ScalarKind::Float }, Dimension::Scalar { .. }) => {...} // floating-point path
(Dimension::Scalar { .. }, Dimension::Scalar { .. }) => {...} // integer path
```

Reply:
I wasn't even aware that matching that way works 😮 I removed the kind, so I went with the `scalar_kind()` function.